dd-sdk-android/src/main/kotlin/com/datadog/android/rum/internal/monitor/DatadogRumMonitor.kt

```diff
@@ -0,0 +1,124 @@
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-2020 Datadog, Inc.
+ */
+
+package com.datadog.android.rum.internal.monitor
+
+import com.datadog.android.core.internal.data.Writer
+import com.datadog.android.core.internal.time.TimeProvider
+import com.datadog.android.core.internal.utils.devLogger
+import com.datadog.android.rum.RumContext
+import com.datadog.android.rum.RumMonitor
+import com.datadog.android.rum.internal.domain.RumEvent
+import com.datadog.android.rum.internal.domain.RumEventData
+import java.util.UUID
+
+internal class DatadogRumMonitor(
+    applicationId: UUID,
+    private val timeProvider: TimeProvider,
+    private val writer: Writer<RumEvent>
+) : RumMonitor {
+
+    private var context: RumContext =
+        RumContext(
+            applicationId,
+            UUID.randomUUID(),
+            null
+        )
+
+    private val activeView = mutableMapOf<Any, RumEvent>()
+
+    // region RumMonitor
+
+    override fun addUserAction(
+        action: String,
+        attributes: Map<String, Any?>
+    ) {
+        val eventData = RumEventData.UserAction(action)
+        val event = RumEvent(
+            context,
+            timeProvider.getServerTimestamp(),
+            eventData,
+            attributes
+        )
+        writer.write(event)
+    }
+
+    override fun startView(
+        key: Any,
+        className: String,
+        attributes: Map<String, Any?>
+    ) {
+        val activeViewId = UUID.randomUUID()
+        context = context.copy(viewId = activeViewId)
+        val eventData = RumEventData.View(className, System.nanoTime())
+        val event = RumEvent(
+            context,
+            timeProvider.getServerTimestamp(),
+            eventData,
+            attributes
+        )
+        activeView[key] = event
+    }
+
+    override fun stopView(
+        key: Any,
+        className: String,
+        attributes: Map<String, Any?>
+    ) {
+        val startedEvent = activeView[key]
+        val startedEventData = startedEvent?.eventData as? RumEventData.View
+
+        if (startedEvent == null) {
+            devLogger.w(
+                "Unable to stop view with key <$key> and class <$className>. " +
+                    "This can mean that the view was not started or already stopped."
+            )
+        } else if (startedEventData == null) {
+            devLogger.e(
+                "Unable to stop view with key <$key> and class <$className>. " +
+                    "The related data was inconsistent."
+            )
+        } else {
+            activeView.remove(key)
+
+            val updatedEvent = startedEvent.copy(
+                eventData = startedEventData.copy(
+                    durationNanoSeconds = System.nanoTime() - startedEventData.durationNanoSeconds,
+                    version = startedEventData.version + 1
```

Comment:
Why do we increment `version`? What is `version` then 🤔?

Reply:
The View is a mutable object, but because updates can be sent asynchronously, each update has a `rum.document_version` field, which needs to be incremented. If the server receives version 3 after version 6, it will keep version 6.